File: haptic/api/python/wia_haptic/device.py
"""
WIA Haptic Standard - Device Abstraction Layer
"""

from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from typing import Optional, List, Dict, Callable, Any, Union
import asyncio

from .types import (
    BodyLocation,
    ActuatorType,
    DeviceType,
    ConnectionState,
    HapticErrorCode,
)
from .patterns import HapticPattern, HapticPrimitive, HapticSequence, SpatialPattern


# ============================================================================
# Data Classes
# ============================================================================

@dataclass
class HapticCapabilities:
    """Device capabilities"""
    actuator_type: ActuatorType
    frequency_range: tuple  # (min, max) in Hz
    locations: List[BodyLocation]
    max_intensity: float = 1.0
    latency: float = 20  # ms
    actuator_count: int = 1
    supports_custom_waveforms: bool = False
    supports_amplitude_modulation: bool = True
    supports_frequency_modulation: bool = False
    battery_level: Optional[int] = None  # 0-100 or None


@dataclass
class UserCalibration:
    """User-specific calibration settings"""
    perception_thresholds: Dict[BodyLocation, float] = field(default_factory=dict)
    comfort_maximum: Dict[BodyLocation, float] = field(default_factory=dict)
    unavailable_locations: List[BodyLocation] = field(default_factory=list)


@dataclass
class DeviceConfig:
    """Device configuration"""
    device_id: Optional[str] = None
    name: Optional[str] = None
    connection_timeout: float = 10000  # ms
    auto_reconnect: bool = True
    intensity_scale: float = 1.0
    calibration: Optional[UserCalibration] = None


@dataclass
class HapticDeviceEvent:
    """Event from haptic device"""
    type: str  # connected, disconnected, error, battery, pattern_complete
    timestamp: float
    data: Any = None


# Type alias for event listeners
HapticEventListener = Callable[[HapticDeviceEvent], None]


# ============================================================================
# Exceptions
# ============================================================================

class HapticError(Exception):
    """Haptic operation error"""

    def __init__(
        self,
        message: str,
        code: HapticErrorCode = HapticErrorCode.UNKNOWN,
        details: Any = None
    ):
        super().__init__(message)
        self.code = code
        self.details = details


# ============================================================================
# Abstract Device Interface
# ============================================================================

class IHapticDevice(ABC):
    """Interface for haptic devices"""

    @property
    @abstractmethod
    def state(self) -> ConnectionState:
        """Current connection state"""
        pass

    @property
    @abstractmethod
    def capabilities(self) -> Optional[HapticCapabilities]:
        """Device capabilities (available after connection)"""
        pass

    @property
    @abstractmethod
    def device_type(self) -> DeviceType:
        """Device type"""
        pass

    @property
    @abstractmethod
    def config(self) -> DeviceConfig:
        """Device configuration"""
        pass

    @abstractmethod
    async def connect(self) -> None:
        """Connect to the haptic device"""
        pass

    @abstractmethod
    async def disconnect(self) -> None:
        """Disconnect from the haptic device"""
        pass

    @abstractmethod
    def is_connected(self) -> bool:
        """Check if device is connected"""
        pass

    @abstractmethod
    def get_capabilities(self) -> HapticCapabilities:
        """Get device capabilities"""
        pass

    @abstractmethod
    def supports_location(self, location: BodyLocation) -> bool:
        """Check if device supports a specific location"""
        pass

    @abstractmethod
    def supports_frequency(self, frequency: float) -> bool:
        """Check if device supports a specific frequency"""
        pass

    @abstractmethod
    async def play(self, pattern: HapticPattern) -> None:
        """Play a haptic pattern"""
        pass

    @abstractmethod
    async def play_sequence(self, patterns: List[HapticPattern]) -> None:
        """Play multiple patterns in sequence"""
        pass

    @abstractmethod
    def stop(self) -> None:
        """Stop all haptic playback"""
        pass

    @abstractmethod
    def pause(self) -> None:
        """Pause current playback"""
        pass

    @abstractmethod
    def resume(self) -> None:
        """Resume paused playback"""
        pass

    @abstractmethod
    def set_intensity(self, location: BodyLocation, intensity: float) -> None:
        """Set intensity for a specific location"""
        pass

    @abstractmethod
    def pulse(
        self,
        location: BodyLocation,
        duration: float,
        intensity: float = 0.5
    ) -> None:
        """Trigger a quick pulse at a location"""
        pass

    @abstractmethod
    def vibrate(
        self,
        location: BodyLocation,
        frequency: float,
        intensity: float
    ) -> None:
        """Set continuous vibration at a location"""
        pass

    @abstractmethod
    def add_event_listener(self, listener: HapticEventListener) -> None:
        """Add event listener"""
        pass

    @abstractmethod
    def remove_event_listener(self, listener: HapticEventListener) -> None:
        """Remove event listener"""
        pass


# ============================================================================
# Base Device Implementation
# ============================================================================

class BaseHapticDevice(IHapticDevice):
    """Base implementation of haptic device"""

    def __init__(
        self,
        device_type: DeviceType,
        config: Optional[DeviceConfig] = None
    ):
        self._device_type = device_type
        self._config = config or DeviceConfig()
        self._state = ConnectionState.DISCONNECTED
        self._capabilities: Optional[HapticCapabilities] = None
        self._listeners: List[HapticEventListener] = []
        self._is_playing = False
        self._is_paused = False

    @property
    def state(self) -> ConnectionState:
        return self._state

    @property
    def capabilities(self) -> Optional[HapticCapabilities]:
        return self._capabilities

    @property
    def device_type(self) -> DeviceType:
        return self._device_type

    @property
    def config(self) -> DeviceConfig:
        return self._config

    def is_connected(self) -> bool:
        return self._state == ConnectionState.CONNECTED

    def get_capabilities(self) -> HapticCapabilities:
        if not self._capabilities:
            raise HapticError("Device not connected", HapticErrorCode.NOT_CONNECTED)
        return self._capabilities

    def supports_location(self, location: BodyLocation) -> bool:
        if not self._capabilities:
            return False
        return location in self._capabilities.locations

    def supports_frequency(self, frequency: float) -> bool:
        if not self._capabilities:
            return False
        min_freq, max_freq = self._capabilities.frequency_range
        return min_freq <= frequency <= max_freq

    async def play_sequence(self, patterns: List[HapticPattern]) -> None:
        for pattern in patterns:
            await self.play(pattern)

    def pause(self) -> None:
        if self._is_playing:
            self._is_paused = True

    def resume(self) -> None:
        self._is_paused = False

    def add_event_listener(self, listener: HapticEventListener) -> None:
        if listener not in self._listeners:
            self._listeners.append(listener)

    def remove_event_listener(self, listener: HapticEventListener) -> None:
        if listener in self._listeners:
            self._listeners.remove(listener)

    def _emit(self, event: HapticDeviceEvent) -> None:
        """Emit event to all listeners"""
        for listener in self._listeners:
            try:
                listener(event)
            except Exception as e:
                print(f"Error in haptic event listener: {e}")

    def _assert_connected(self) -> None:
        """Assert device is connected"""
        if not self.is_connected():
            raise HapticError("Device not connected", HapticErrorCode.NOT_CONNECTED)

    def _scale_intensity(self, intensity: float) -> float:
        """Scale intensity by config scale factor"""
        scale = self._config.intensity_scale
        return max(0.0, min(1.0, intensity * scale))

    def _apply_calibration(
        self,
        location: BodyLocation,
        intensity: float
    ) -> float:
        """Apply user calibration to intensity"""
        calibration = self._config.calibration
        if not calibration:
            return intensity

        # Check if location is unavailable
        if location in calibration.unavailable_locations:
            return 0.0

        # Apply perception threshold
        threshold = calibration.perception_thresholds.get(location, 0.0)
        if intensity < threshold:
            return 0.0

        # Apply comfort maximum
        maximum = calibration.comfort_maximum.get(location, 1.0)
        return min(intensity, maximum)

    # Abstract methods to be implemented by subclasses
    @abstractmethod
    async def connect(self) -> None:
        pass

    @abstractmethod
    async def disconnect(self) -> None:
        pass

    @abstractmethod
    async def play(self, pattern: HapticPattern) -> None:
        pass

    @abstractmethod
    def stop(self) -> None:
        pass

    @abstractmethod
    def set_intensity(self, location: BodyLocation, intensity: float) -> None:
        pass

    @abstractmethod
    def pulse(
        self,
        location: BodyLocation,
        duration: float,
        intensity: float = 0.5
    ) -> None:
        pass

    @abstractmethod
    def vibrate(
        self,
        location: BodyLocation,
        frequency: float,
        intensity: float
    ) -> None:
        pass


# ============================================================================
# Device Manager
# ============================================================================

class HapticDeviceManager:
    """Manages multiple haptic devices"""

    def __init__(self):
        self._devices: Dict[str, IHapticDevice] = {}
        self._default_device: Optional[IHapticDevice] = None

    def register(self, device_id: str, device: IHapticDevice) -> None:
        """Register a device"""
        self._devices[device_id] = device
        if self._default_device is None:
            self._default_device = device

    def unregister(self, device_id: str) -> None:
        """Unregister a device"""
        device = self._devices.get(device_id)
        if device:
            if device.is_connected():
                asyncio.create_task(device.disconnect())
            del self._devices[device_id]
            if self._default_device == device:
                self._default_device = next(iter(self._devices.values()), None)

    def get(self, device_id: str) -> Optional[IHapticDevice]:
        """Get device by ID"""
        return self._devices.get(device_id)

    def get_default(self) -> Optional[IHapticDevice]:
        """Get default device"""
        return self._default_device

    def set_default(self, device_id: str) -> None:
        """Set default device"""
        device = self._devices.get(device_id)
        if device:
            self._default_device = device

    def get_all(self) -> List[IHapticDevice]:
        """Get all registered devices"""
        return list(self._devices.values())

    def get_connected(self) -> List[IHapticDevice]:
        """Get all connected devices"""
        return [d for d in self._devices.values() if d.is_connected()]

    async def connect_all(self) -> None:
        """Connect all devices"""
        tasks = [d.connect() for d in self._devices.values()]
        await asyncio.gather(*tasks, return_exceptions=True)

    async def disconnect_all(self) -> None:
        """Disconnect all devices"""
        tasks = [d.disconnect() for d in self.get_connected()]
        await asyncio.gather(*tasks, return_exceptions=True)

    async def play_on_all(self, pattern: HapticPattern) -> None:
        """Play pattern on all connected devices"""
        tasks = [d.play(pattern) for d in self.get_connected()]
        await asyncio.gather(*tasks, return_exceptions=True)

    def stop_all(self) -> None:
        """Stop all devices"""
        for device in self.get_connected():
            device.stop()


# Singleton instance
haptic_manager = HapticDeviceManager()

File: bionic-eye/haptic/api/typescript/src/device.ts
/**
 * WIA Haptic Standard - Device Abstraction Layer
 * @version 1.0.0
 */

import {
  BodyLocation,
  ActuatorType,
  DeviceType,
  ConnectionState,
  HapticDeviceEvent,
  HapticEventListener,
} from './types';

import {
  HapticPattern,
  HapticPrimitive,
  HapticSequence,
  SpatialPattern,
} from './patterns';

// ============================================================================
// Device Capabilities
// ============================================================================

export interface HapticCapabilities {
  /** Type of haptic actuator */
  actuatorType: ActuatorType;

  /** Supported frequency range in Hz */
  frequencyRange: {
    min: number;
    max: number;
  };

  /** Available body locations */
  locations: BodyLocation[];

  /** Maximum supported intensity (0.0-1.0) */
  maxIntensity: number;

  /** Response latency in milliseconds */
  latency: number;

  /** Number of independent actuators */
  actuatorCount: number;

  /** Whether device supports custom waveforms */
  supportsCustomWaveforms: boolean;

  /** Whether device supports amplitude modulation */
  supportsAmplitudeModulation: boolean;

  /** Whether device supports frequency modulation */
  supportsFrequencyModulation: boolean;

  /** Battery level (0-100) or null if not available */
  batteryLevel: number | null;
}

// ============================================================================
// Device Configuration
// ============================================================================

export interface DeviceConfig {
  /** Device identifier */
  deviceId?: string;

  /** Device name for display */
  name?: string;

  /** Connection timeout in ms */
  connectionTimeout?: number;

  /** Auto-reconnect on disconnect */
  autoReconnect?: boolean;

  /** Global intensity scaling (0.0-2.0) */
  intensityScale?: number;

  /** User-specific calibration */
  calibration?: UserCalibration;
}

export interface UserCalibration {
  /** Per-location intensity thresholds */
  perceptionThresholds?: Partial<Record<BodyLocation, number>>;

  /** Per-location maximum comfort levels */
  comfortMaximum?: Partial<Record<BodyLocation, number>>;

  /** Locations to avoid */
  unavailableLocations?: BodyLocation[];
}

// ============================================================================
// Haptic Device Interface
// ============================================================================

export interface IHapticDevice {
  // -------------------------------------------------------------------------
  // Properties
  // -------------------------------------------------------------------------

  /** Current connection state */
  readonly state: ConnectionState;

  /** Device capabilities (available after connection) */
  readonly capabilities: HapticCapabilities | null;

  /** Device type */
  readonly deviceType: DeviceType;

  /** Device configuration */
  readonly config: DeviceConfig;

  // -------------------------------------------------------------------------
  // Connection Management
  // -------------------------------------------------------------------------

  /**
   * Connect to the haptic device
   * @throws HapticError on connection failure
   */
  connect(): Promise<void>;

  /**
   * Disconnect from the haptic device
   */
  disconnect(): Promise<void>;

  /**
   * Check if device is connected
   */
  isConnected(): boolean;

  // -------------------------------------------------------------------------
  // Capability Query
  // -------------------------------------------------------------------------

  /**
   * Get device capabilities
   * @throws HapticError if not connected
   */
  getCapabilities(): HapticCapabilities;

  /**
   * Check if device supports a specific location
   */
  supportsLocation(location: BodyLocation): boolean;

  /**
   * Check if device supports a specific frequency
   */
  supportsFrequency(frequency: number): boolean;

  // -------------------------------------------------------------------------
  // Pattern Playback
  // -------------------------------------------------------------------------

  /**
   * Play a haptic pattern
   * @param pattern - Primitive, sequence, or spatial pattern
   */
  play(pattern: HapticPattern): Promise<void>;

  /**
   * Play multiple patterns in sequence
   * @param patterns - Array of patterns to play
   */
  playSequence(patterns: HapticPattern[]): Promise<void>;

  /**
   * Stop all haptic playback
   */
  stop(): void;

  /**
   * Pause current playback
   */
  pause(): void;

  /**
   * Resume paused playback
   */
  resume(): void;

  // -------------------------------------------------------------------------
  // Real-time Control
  // -------------------------------------------------------------------------

  /**
   * Set intensity for a specific location
   * @param location - Body location
   * @param intensity - Intensity value (0.0-1.0)
   */
  setIntensity(location: BodyLocation, intensity: number): void;

  /**
   * Trigger a quick pulse at a location
   * @param location - Body location
   * @param duration - Pulse duration in ms
   * @param intensity - Optional intensity (0.0-1.0)
   */
  pulse(location: BodyLocation, duration: number, intensity?: number): void;

  /**
   * Set continuous vibration at a location
   * @param location - Body location
   * @param frequency - Frequency in Hz
   * @param intensity - Intensity (0.0-1.0)
   */
  vibrate(location: BodyLocation, frequency: number, intensity: number): void;

  // -------------------------------------------------------------------------
  // Event Handling
  // -------------------------------------------------------------------------

  /**
   * Add event listener
   */
  addEventListener(listener: HapticEventListener): void;

  /**
   * Remove event listener
   */
  removeEventListener(listener: HapticEventListener): void;
}

// ============================================================================
// Haptic Error
// ============================================================================

export class HapticError extends Error {
  constructor(
    message: string,
    public readonly code: HapticErrorCode,
    public readonly details?: unknown
  ) {
    super(message);
    this.name = 'HapticError';
  }
}

export type HapticErrorCode =
  | 'CONNECTION_FAILED'
  | 'CONNECTION_TIMEOUT'
  | 'DEVICE_NOT_FOUND'
  | 'NOT_CONNECTED'
  | 'UNSUPPORTED_FEATURE'
  | 'INVALID_PATTERN'
  | 'PLAYBACK_ERROR'
  | 'PERMISSION_DENIED'
  | 'BLUETOOTH_DISABLED'
  | 'UNKNOWN';

// ============================================================================
// Abstract Base Device
// ============================================================================

export abstract class BaseHapticDevice implements IHapticDevice {
  protected _state: ConnectionState = 'disconnected';
  protected _capabilities: HapticCapabilities | null = null;
  protected _config: DeviceConfig;
  protected _listeners: Set<HapticEventListener> = new Set();
  protected _isPlaying: boolean = false;
  protected _isPaused: boolean = false;

  constructor(
    public readonly deviceType: DeviceType,
    config: DeviceConfig = {}
  ) {
    this._config = {
      connectionTimeout: 10000,
      autoReconnect: true,
      intensityScale: 1.0,
      ...config,
    };
  }

  // -------------------------------------------------------------------------
  // Properties
  // -------------------------------------------------------------------------

  get state(): ConnectionState {
    return this._state;
  }

  get capabilities(): HapticCapabilities | null {
    return this._capabilities;
  }

  get config(): DeviceConfig {
    return this._config;
  }

  // -------------------------------------------------------------------------
  // Connection (to be implemented by adapters)
  // -------------------------------------------------------------------------

  abstract connect(): Promise<void>;
  abstract disconnect(): Promise<void>;

  isConnected(): boolean {
    return this._state === 'connected';
  }

  // -------------------------------------------------------------------------
  // Capabilities
  // -------------------------------------------------------------------------

  getCapabilities(): HapticCapabilities {
    if (!this._capabilities) {
      throw new HapticError(
        'Device not connected',
        'NOT_CONNECTED'
      );
    }
    return this._capabilities;
  }

  supportsLocation(location: BodyLocation): boolean {
    return this._capabilities?.locations.includes(location) ?? false;
  }

  supportsFrequency(frequency: number): boolean {
    if (!this._capabilities) return false;
    const { min, max } = this._capabilities.frequencyRange;
    return frequency >= min && frequency <= max;
  }

  // -------------------------------------------------------------------------
  // Playback (to be implemented by adapters)
  // -------------------------------------------------------------------------

  abstract play(pattern: HapticPattern): Promise<void>;

  async playSequence(patterns: HapticPattern[]): Promise<void> {
    for (const pattern of patterns) {
      await this.play(pattern);
    }
  }

  abstract stop(): void;

  pause(): void {
    if (this._isPlaying) {
      this._isPaused = true;
    }
  }

  resume(): void {
    this._isPaused = false;
  }

  // -------------------------------------------------------------------------
  // Real-time Control (to be implemented by adapters)
  // -------------------------------------------------------------------------

  abstract setIntensity(location: BodyLocation, intensity: number): void;
  abstract pulse(location: BodyLocation, duration: number, intensity?: number): void;
  abstract vibrate(location: BodyLocation, frequency: number, intensity: number): void;

  // -------------------------------------------------------------------------
  // Event Handling
  // -------------------------------------------------------------------------

  addEventListener(listener: HapticEventListener): void {
    this._listeners.add(listener);
  }

  removeEventListener(listener: HapticEventListener): void {
    this._listeners.delete(listener);
  }

  protected emit(event: HapticDeviceEvent): void {
    for (const listener of this._listeners) {
      try {
        listener(event);
      } catch (e) {
        console.error('Error in haptic event listener:', e);
      }
    }
  }

  // -------------------------------------------------------------------------
  // Utility Methods
  // -------------------------------------------------------------------------

  protected assertConnected(): void {
    if (!this.isConnected()) {
      throw new HapticError('Device not connected', 'NOT_CONNECTED');
    }
  }

  protected scaleIntensity(intensity: number): number {
    const scale = this._config.intensityScale ?? 1.0;
    return Math.min(1.0, Math.max(0.0, intensity * scale));
  }

  protected applyCalibration(
    location: BodyLocation,
    intensity: number
  ): number {
    const calibration = this._config.calibration;
    if (!calibration) return intensity;

    // Check if location is unavailable
    if (calibration.unavailableLocations?.includes(location)) {
      return 0;
    }

    // Apply perception threshold
    const threshold = calibration.perceptionThresholds?.[location] ?? 0;
    if (intensity < threshold) {
      return 0;
    }

    // Apply comfort maximum
    const maximum = calibration.comfortMaximum?.[location] ?? 1.0;
    return Math.min(intensity, maximum);
  }
}

// ============================================================================
// Device Manager
// ============================================================================

export class HapticDeviceManager {
  private devices: Map<string, IHapticDevice> = new Map();
  private defaultDevice: IHapticDevice | null = null;

  /**
   * Register a device
   */
  register(id: string, device: IHapticDevice): void {
    this.devices.set(id, device);
    if (!this.defaultDevice) {
      this.defaultDevice = device;
    }
  }

  /**
   * Unregister a device
   */
  unregister(id: string): void {
    const device = this.devices.get(id);
    if (device) {
      if (device.isConnected()) {
        device.disconnect();
      }
      this.devices.delete(id);
      if (this.defaultDevice === device) {
        this.defaultDevice = this.devices.values().next().value ?? null;
      }
    }
  }

  /**
   * Get device by ID
   */
  get(id: string): IHapticDevice | undefined {
    return this.devices.get(id);
  }

  /**
   * Get default device
   */
  getDefault(): IHapticDevice | null {
    return this.defaultDevice;
  }

  /**
   * Set default device
   */
  setDefault(id: string): void {
    const device = this.devices.get(id);
    if (device) {
      this.defaultDevice = device;
    }
  }

  /**
   * Get all registered devices
   */
  getAll(): IHapticDevice[] {
    return Array.from(this.devices.values());
  }

  /**
   * Get all connected devices
   */
  getConnected(): IHapticDevice[] {
    return this.getAll().filter(d => d.isConnected());
  }

  /**
   * Connect all devices
   */
  async connectAll(): Promise<void> {
    await Promise.all(
      this.getAll().map(d => d.connect().catch(e => console.error(e)))
    );
  }

  /**
   * Disconnect all devices
   */
  async disconnectAll(): Promise<void> {
    await Promise.all(
      this.getConnected().map(d => d.disconnect())
    );
  }

  /**
   * Play pattern on all connected devices
   */
  async playOnAll(pattern: HapticPattern): Promise<void> {
    await Promise.all(
      this.getConnected().map(d => d.play(pattern))
    );
  }

  /**
   * Stop all devices
   */
  stopAll(): void {
    for (const device of this.getConnected()) {
      device.stop();
    }
  }
}

// ============================================================================
// Singleton Manager Instance
// ============================================================================

export const hapticManager = new HapticDeviceManager();
